editoast/src/views/mod.rs

```
@@ -200,6 +201,78 @@ impl Authentication {
+                        );
+                        Ok(false)
+                    }
+                    _ => Ok(true),
+                }
+            }
+        }
+    }
+
+    /// Checks if the issuer of the request has the can_read privilege on the provided infra. Always returns `false` if the
+    /// request is unauthenticated.
+    async fn check_infra_can_write(&self, infra_id: i64) -> Result<bool, AuthorizerError> {
+        match self {
+            Authentication::SkipAuthorization { .. } => Ok(true),
+            Authentication::Unauthenticated => Ok(false),
+            Authentication::Authenticated(authorizer) => {
+                match authorizer.authorize_infra_write(infra_id).await? {
+                    Authorization::Denied { reason } => {
+                        tracing::debug!(
+                            "User is not authorized to read infra {infra_id}: {reason}"
+                        );
+                        Ok(false)
+                    }
+                    _ => Ok(true),
+                }
+            }
+        }
+    }
+
+    /// Checks if the issuer of the request has the can_delete privilege on the provided infra. Always returns `false` if the
+    /// request is unauthenticated.
+    async fn check_infra_can_delete(&self, infra_id: i64) -> Result<bool, AuthorizerError> {
+        match self {
+            Authentication::SkipAuthorization { .. } => Ok(true),
+            Authentication::Unauthenticated => Ok(false),
+            Authentication::Authenticated(authorizer) => {
+                match authorizer.authorize_infra_delete(infra_id).await? {
+                    Authorization::Denied { reason } => {
+                        tracing::debug!(
+                            "User is not authorized to delete infra {infra_id}: {reason}"
+                        );
+                        Ok(false)
+                    }
+                    _ => Ok(true),
+                }
+            }
+        }
+    }
```

These functions can be easily factorize and could delegate the task of converting the unauthorization into an error to the `Authorization` object.

We could have something in the lines of:

```rust
async fn authorize<E: Into<AuthorizationError>>(self, f: impl AsyncFnOnce(Authorizer<PgAuthDriver>) -> Result<Authorization<()>, E>) -> Result<(), AuthorizationError> {
	match self {
		Authenticated(authorizer) => f(authorizer).await?.allowed().map_err(|Unauthorized| AuthorizationError::Forbidden)
		...
	}
}
```

so we can just write in each endpoint:

```rust
auth.authorize(async |authorizer| authorizer.authorize_infra_read(...).await).await?;
```

That brings us a one-liner and hides error management.

wdyt?

---

IMHO, all the `check` functions (check_roles, check_infra_can_read, ...) should return a Forbidden or Unauthenticated error, so we can have  a one line call in views. 

But making an `authorize` function that takes in argument a function with the authorizer can help to factorize the code.  We need to be sure that the arms `Authentication::SkipAuthorization` returns true, and `Unauthenticated` returns false (we don't have the authorizer when skipping auth). 
So I would rename this function `checkAuthorize` (for me a check returns true/false or Ok/Error).

On the other side, devs should know the API of the authorizer and not only the one from Auth, but I think it's a acceptable tradeoff, to avoid code duplication in checks function.

---

I did the job for the infra_check, but we also need to do it check role. 
I did not do it, otherwise this PR will modify all views and so will be huge.

---

Thanks for making the change!

* Agreed about the `check` prefix since we only accept `Authorization<()>`.  (I have in mind that we could later make that function generic over some `T`, in which case I'm not so sure about the prefix, but that's for later.  Better being correct now.)  Nitpick: I believe `check_authorized` would be more correct.
* I think everyone programming with permissions should be comfortable using the `Authorizer` at least.  In my mind, the `Authentication` is just a way to force us to deal with the different authentication modes of editoast.  It shouldn't deal with permissions (or be at most a *very thin* wrapper only where it makes sense).
* The long-term plan we have for roles is to have a middleware in an `axum` router perform the role-checking.  This way we can configure it for entire path sub-trees and remove all explicit role checks in endpoints.